boa/src/syntax/lexer/comment.rs

```diff
@@ -0,0 +1,66 @@
+use super::{Cursor, Error, Tokenizer};
+use crate::syntax::ast::{Position, Span};
+use crate::syntax::lexer::{Token, TokenKind};
+use std::io::Read;
+
+pub(super) struct SingleLineComment;
+
+/// Lexes a single line comment.
+///
+/// Assumes that the initial '//' is already consumed.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-comments
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar
+impl<R> Tokenizer<R> for SingleLineComment {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        // Skip either to the end of the line or to the end of the input
+        while let Some(ch) = cursor.next()? {
+            if ch == '\n' {
+                break;
+            }
+        }
+        Ok(Token::new(
+            TokenKind::Comment,
+            Span::new(start_pos, cursor.pos()),
+        ))
+    }
+}
+
+/// Lexes a block (multi-line) comment.
+///
+/// Assumes that the initial '/*' is already consumed.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-comments
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar
+pub(super) struct BlockComment;
+impl<R> Tokenizer<R> for BlockComment {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        loop {
+            if let Some(ch) = cursor.next()? {
+                if ch == '*' && cursor.next_is('/')? {
+                    break;
+                }
+            } else {
+                return Err(Error::syntax("unterminated multiline comment"));
+            }
+        }
+        Ok(Token::new(
+            TokenKind::Comment,
+            Span::new(start_pos, cursor.pos()),
+        ))
+    }
+}
```

Comment:
As per the spec:
> Comments behave like white space and are discarded except that, if a MultiLineComment contains a line terminator code point, then the entire comment is considered to be a LineTerminator for purposes of parsing by the syntactic grammar.

So, most of the time, this just works, but there is a particular case where this is not spec compliant:

```javascript
let a = 10 /* Test
Multiline
Comment
*/ let b = 10;
```

This should be equivalent to:
```javascript
let a = 10
let b = 10;
```

But once again, it happens the same as in single line comments, where the new line doesn't get inserted, and therefore we don't have an automatic semicolon insertion.

Note that this new line should only be inserted when the comment has one new line character in it `\n`, so this should not insert a new line:
```javascript
let a = 10; /* Test comment */ let b = 10;
```

So, basically, we need to check if we find a `\n` character inside the multiline comment. If we do, then the returned token should be a `LineTerminator`, if not, then we return a `Comment`, that will later be discarded.

Once again, I would add this as a test case.

Reply:
Done (added 2 test cases in parser/tests)